File: app/src/main/java/opencrowd/hgc/hgcwallet/database/DBHelper.java
package opencrowd.hgc.hgcwallet.database;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.hederahashgraph.api.proto.java.Transaction;
import com.hederahashgraph.api.proto.java.TransactionID;
import com.hederahashgraph.api.proto.java.TransactionRecord;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import opencrowd.hgc.hgcwallet.App;
import opencrowd.hgc.hgcwallet.hapi.APIRequestBuilder;
import opencrowd.hgc.hgcwallet.database.account.Account;
import opencrowd.hgc.hgcwallet.database.account.AccountDao;
import opencrowd.hgc.hgcwallet.database.contact.Contact;
import opencrowd.hgc.hgcwallet.database.contact.ContactDao;
import opencrowd.hgc.hgcwallet.database.node.Node;
import opencrowd.hgc.hgcwallet.database.node.NodeDao;
import opencrowd.hgc.hgcwallet.database.request.PayRequest;
import opencrowd.hgc.hgcwallet.database.request.PayRequestDao;
import opencrowd.hgc.hgcwallet.database.transaction.TxnRecord;
import opencrowd.hgc.hgcwallet.database.transaction.TxnRecordDao;
import opencrowd.hgc.hgcwallet.database.wallet.Wallet;
import opencrowd.hgc.hgcwallet.database.wallet.WalletDao;
import opencrowd.hgc.hgcwallet.modals.HGCKeyType;
import opencrowd.hgc.hgcwallet.modals.HGCAccountID;

public class DBHelper {

    public static void createMasterWallet(@NonNull HGCKeyType type) {
        WalletDao walletDao = App.instance.database.walletDao();
        Wallet wallet = null;
        List<Wallet> allWallets = walletDao.getAllWallets();
        if (allWallets != null && allWallets.size() > 0) {
            wallet = allWallets.get(0);
            // something is wrong
        } else {
            long walletId = walletDao.insert(Wallet.Companion.createWallet(type));
            createNewAccount("Default Account");
        }
    }

    @Nullable
    public static Wallet getMasterWallet() {
        WalletDao walletDao = App.instance.database.walletDao();
        Wallet wallet = null;
        List<Wallet> allWallets = walletDao.getAllWallets();
        if (allWallets != null && allWallets.size() > 0) {
            wallet = allWallets.get(0);
        }
        return wallet;
    }

    @NonNull
    public static List<Account> getAllAccounts() {
        Wallet wallet = getMasterWallet();
        AccountDao accountDao = App.instance.database.accountDao();
        List<Account> accounts = accountDao.findAccountForWallet(wallet.getWalletId());
        return accounts;
    }

    @NonNull
    public static Account createNewAccount(@NonNull String accountName) {
        WalletDao walletDao = App.instance.database.walletDao();
        AccountDao accountDao = App.instance.database.accountDao();

        Wallet wallet = getMasterWallet();
        Account account = null;
        if (wallet != null) {
            long index = wallet.getTotalAccounts();
            accountDao.insert(Account.Companion.createAccount(wallet.getWalletId(), wallet.getKeyType(), index, accountName));
            wallet.setTotalAccounts(wallet.getTotalAccounts() +1);
            walletDao.update(wallet);
            account = accountDao.findAccountForIndex(index).get(0);
        }
        return account;
    }

    public static void saveAccount(@NonNull Account account) {
        AccountDao accountDao = App.instance.database.accountDao();
        accountDao.update(account);
    }

    @Nullable
    public static List<Contact> getAllContacts() {
        ContactDao contactDao = App.instance.database.contactDao();
        return contactDao.getAllContacts();
    }

    @Nullable
    public static Contact getContact(@NonNull String accountID) {
        ContactDao contactDao = App.instance.database.contactDao();
        List<Contact> contacts = contactDao.findContactForAccountId(accountID);
        if (contacts != null && !contacts.isEmpty()) {
            return contacts.get(0);
        }
        return null;
    }

    public static void createContact(@NonNull HGCAccountID accountID, @Nullable String name, boolean isVerified) {
        if (accountID == null) return;
        if (name == null) name = "";

        ContactDao contactDao = App.instance.database.contactDao();
        Contact contact = getContact(accountID.stringRepresentation());
        if (contact == null) {
            contact = new Contact(accountID.stringRepresentation(), name, isVerified);
            contactDao.insert(contact);
        } else {
            if (name != null && contact.getName() != null) {
                if (name.equals(contact.getName()) && isVerified) {
                    contact.setVerified(true);
                } else {
                    contact.setVerified(isVerified);
                }
                contact.setName(name);
                saveContact(contact);
            }
        }
    }

    public static void saveContact(@NonNull Contact contact) {
        ContactDao contactDao = App.instance.database.contactDao();
        contactDao.update(contact);
    }

    @Nullable
    public static List<PayRequest> getAllRequests() {
        PayRequestDao payRequestDao = App.instance.database.payRequestDao();
        return payRequestDao.getAllPayRequests();
    }

    public static PayRequest getPayRequest(@NonNull HGCAccountID accountID, long amount, @Nullable String name, @Nullable String notes) {
        PayRequestDao payRequestDao = App.instance.database.payRequestDao();
        List<PayRequest> payRequests = payRequestDao.findPayRequest(accountID.stringRepresentation(), name, amount, notes);
        if (payRequests != null && !payRequests.isEmpty()) {
            return payRequests.get(0);
        }
        return null;
    }

    public static void createPayRequest(@NonNull HGCAccountID accountID, long amount, @Nullable String name, @Nullable String notes) {
        PayRequestDao payRequestDao = App.instance.database.payRequestDao();
        PayRequest payRequest = getPayRequest(accountID, amount, name, notes);
        if (payRequest == null) {
            payRequest = new PayRequest(0, accountID.stringRepresentation(), name, notes, amount, new Date());
            payRequestDao.insert(payRequest);

        } else {
            payRequest.setImportDate(new Date());
            savePayRequest(payRequest);
        }
    }

    public static void savePayRequest(@NonNull PayRequest request) {
        PayRequestDao payRequestDao = App.instance.database.payRequestDao();
        payRequestDao.update(request);
    }

    public static void deletePayRequest(@NonNull PayRequest request) {
        PayRequestDao payRequestDao = App.instance.database.payRequestDao();
        payRequestDao.delete(request);
    }

    public static TxnRecord createTransaction(@NonNull Transaction txn, HGCAccountID fromAccount) {
        TxnRecordDao dao = App.instance.database.txnRecordDao();
        TxnRecord record = getTransaction(APIRequestBuilder.getTxnBody(txn).getTransactionID());
        if (record == null) {
            record = new TxnRecord(APIRequestBuilder.getTxnBody(txn).getTransactionID());
            record.setTxn(txn.toByteArray());
            record.setCreatedDate( new Date());
            record.setFromAccId(fromAccount.stringRepresentation());
            dao.insert(record);
            return dao.findRecordForTxnId(Converters.transactionIdToString(record.getTxnId())).get(0);
        } else {
            return record;
        }
    }

    public static TxnRecord createTransaction(@NonNull TransactionRecord transactionRecord, HGCAccountID fromAccount) {
        TxnRecordDao dao = App.instance.database.txnRecordDao();
        TxnRecord record = getTransaction(transactionRecord.getTransactionID());
        if (record == null) {
            record = new TxnRecord(transactionRecord.getTransactionID());
            record.setRecord( transactionRecord.toByteArray());
            record.setCreatedDate(new Date());
            record.setFromAccId(fromAccount.stringRepresentation());
            dao.insert(record);
            return dao.findRecordForTxnId(Converters.transactionIdToString(record.getTxnId())).get(0);
        } else {
            return record;
        }
    }

    public static TxnRecord getTransaction(@NonNull TransactionID txnId) {
        String txnIdStr = Converters.transactionIdToString(txnId);
        TxnRecordDao dao = App.instance.database.txnRecordDao();
        List<TxnRecord> list = dao.findRecordForTxnId(txnIdStr);
        if (list != null && !list.isEmpty()) {
            return list.get(0);
        }
        return null;
    }

    public static void updateTransaction(TxnRecord record) {
        TxnRecordDao dao = App.instance.database.txnRecordDao();
        dao.update(record);
    }

    public static List<TxnRecord> getAllTxnRecord(@Nullable Account fromAccount) {

        TxnRecordDao dao = App.instance.database.txnRecordDao();
        List<TxnRecord> records = null;
        if (fromAccount == null) {
            records = dao.getAllRecords();
        } else {
            if (fromAccount.accountID() != null) {
                records = dao.findRecordForAccountId(fromAccount.accountID().stringRepresentation());
            } else {
                records = new ArrayList<>();
            }
        }

        if (records != null && !records.isEmpty()) {
            List<Account> accounts = getAllAccounts();
            for (TxnRecord record : records) {
                record.parseProperties();
                if (record.getFromAccId() != null) {
                    Account account = getAccount(accounts, record.getFromAccId());
                    if (account != null) {
                        record.setFromAccount( account.getContact());
                    } else {
                        Contact contact = getContact(record.getFromAccId());
                        if (contact != null) {
                            record.setFromAccount( contact);
                        }
                    }
                }

                if (record.getToAccountId() != null) {
                    Account account = getAccount(accounts, record.getToAccountId());
                    if (account != null) {
                        record.setToAccount(account.getContact());
                    } else {
                        Contact contact = getContact(record.getToAccountId());
                        if (contact != null) {
                            record.setToAccount( contact);
                        }
                    }
                    if (fromAccount != null && fromAccount.accountID() != null) {
                        if (record.getToAccountId().equals(fromAccount.accountID().stringRepresentation())) {
                            record.setPositive(true);
                        }
                    }
                }
            }
        }


        return records;
    }

    public static void deleteTxnRecord(TxnRecord record) {
        TxnRecordDao dao = App.instance.database.txnRecordDao();
        dao.delete(record);
    }

    private static Account getAccount(List<Account> accounts, String accountId) {
        HGCAccountID accountID = HGCAccountID.fromString(accountId);
        if (accountID != null) {
            for (Account account : accounts) {
                HGCAccountID accId = account.accountID();
                if (accId != null && accId.stringRepresentation().equals(accountID.stringRepresentation())) {
                    return account;
                }
            }
        }
        return null;
    }

    public static Node getNode(@NonNull String host) {
        NodeDao dao = App.instance.database.nodeDao();
        List<Node> list = dao.findNodeForHost(host);
        if (list != null && !list.isEmpty()) {
            return list.get(0);
        }
        return null;
    }

    public static void createNode(@NonNull Node node) {
        NodeDao dao = App.instance.database.nodeDao();
        Node record = getNode(node.getHost());
        if (record == null) {
            dao.insert(node);
        }
    }

    public static void updateNode(@NonNull Node node) {
        NodeDao dao = App.instance.database.nodeDao();
        dao.update(node);
    }

    public static List<Node> getAllNodes(Boolean activeOnly) {
        NodeDao dao = App.instance.database.nodeDao();
        if (activeOnly)
            return dao.findActiveNodes();
        else
            return dao.getAllNodes();
    }
}

File: app/src/main/java/hedera/hgc/hgcwallet/database/DBHelper.kt
/*
 *
 *  Copyright 2019 Hedera Hashgraph LLC
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package hedera.hgc.hgcwallet.database


import com.hederahashgraph.api.proto.java.Transaction
import com.hederahashgraph.api.proto.java.TransactionID
import com.hederahashgraph.api.proto.java.TransactionRecord
import hedera.hgc.hgcwallet.App
import hedera.hgc.hgcwallet.database.account.Account
import hedera.hgc.hgcwallet.database.account.AccountType
import hedera.hgc.hgcwallet.database.contact.Contact
import hedera.hgc.hgcwallet.database.node.Node
import hedera.hgc.hgcwallet.database.request.PayRequest
import hedera.hgc.hgcwallet.database.transaction.TxnRecord
import hedera.hgc.hgcwallet.database.wallet.Wallet
import hedera.hgc.hgcwallet.database.wallet.WalletDao
import hedera.hgc.hgcwallet.hapi.getTxnBody
import hedera.hgc.hgcwallet.modals.HGCKeyType
import hedera.hgc.hgcwallet.modals.HGCAccountID
import hedera.hgc.hgcwallet.modals.KeyDerivation
import io.reactivex.Flowable
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

object DBHelper {

    fun createMasterWallet(keyDerivation: KeyDerivation) {
        val walletDao = App.instance.database!!.walletDao()
        var wallet: Wallet? = null
        val allWallets = walletDao.getAllWallets()
        if (allWallets != null && allWallets.size > 0) {
            wallet = allWallets[0]
            // something is wrong
        } else {
            val walletId = walletDao.insert(Wallet.createWallet(HGCKeyType.ED25519, keyDerivation))
            createNewAccount("Default Account")
        }
    }

    fun getMasterWallet(): Wallet? {
        val walletDao = App.instance.database!!.walletDao()
        var wallet: Wallet? = null
        val allWallets = walletDao.getAllWallets()
        if (allWallets != null && allWallets.size > 0) {
            wallet = allWallets[0]
        }
        return wallet
    }

    fun getAllAccounts(accountType: AccountType = AccountType.AUTO): List<Account> {
        val wallet = getMasterWallet()
        val accountDao = App.instance.database!!.accountDao()
        return accountDao.findAccountForWallet(wallet!!.walletId, accountType.value)
    }

    fun getAllAccountsFlowable(accountType: AccountType = AccountType.AUTO): Flowable<List<Account>> {
        val wallet = getMasterWallet()
        val accountDao = App.instance.database!!.accountDao()
        return accountDao.findAccountsForWalletFlowable(wallet!!.walletId, accountType.value)
    }

    fun createNewAccount(accountName: String): Account? {
        val walletDao = App.instance.database!!.walletDao()
        val accountDao = App.instance.database!!.accountDao()

        val wallet = getMasterWallet()
        var account: Account? = null
        if (wallet != null) {
            val index = wallet.totalAccounts
            val uid = accountDao.insert(Account.createAccount(wallet.walletId, wallet.keyType, index, accountName))
            wallet.totalAccounts = wallet.totalAccounts + 1
            walletDao.update(wallet)
            account = accountDao.findAccountForUID(uid)
        }
        return account
    }

    fun createExternalAccount(accountName: String, accountID: HGCAccountID): Account? {
        val walletDao = App.instance.database!!.walletDao()
        val accountDao = App.instance.database!!.accountDao()

        val wallet = getMasterWallet()
        var account: Account? = null
        if (wallet != null) {
            val uid = accountDao.insert(Account.createExternalAccount(wallet.walletId, accountName, accountID))
            account = accountDao.findAccountForUID(uid)
        }
        return account
    }

    fun saveAccount(account: Account) {
        val accountDao = App.instance.database!!.accountDao()
        accountDao.update(account)
    }

    fun deleteAccount(account: Account) {
        val accountDao = App.instance.database!!.accountDao()
        accountDao.delete(account)
    }

    fun getAllContacts(): List<Contact>? {
        val contactDao = App.instance.database!!.contactDao()
        return contactDao.getAllContacts()
    }

    fun getContact(accountID: String): Contact? {
        val contactDao = App.instance.database!!.contactDao()
        val contacts = contactDao.findContactForAccountId(accountID)
        return if (contacts != null && !contacts.isEmpty()) {
            contacts[0]
        } else null
    }

    fun createContact(accountID: HGCAccountID, name: String?, host: String, isVerified: Boolean) {
        var name = name
        if (accountID == null) return
        if (name == null) name = ""

        val contactDao = App.instance.database!!.contactDao()
        var contact = getContact(accountID.stringRepresentation())
        val converter = Converters()
        if (contact == null) {
            var metaData = ""
            if (!host.isEmpty()) {
                val metaDataMap = HashMap<String, String>()
                metaDataMap["host"] = host
                metaData = converter.fromStringMap(metaDataMap)
            }
            contact = Contact(accountID.stringRepresentation(), name, isVerified, metaData)
            contactDao.insert(contact)
        } else {
            if (!host.isEmpty()) {
                val metaDataMap: MutableMap<String, String>
                if (contact.metaData.isEmpty())
                    metaDataMap = HashMap()
                else
                    metaDataMap = converter.fromString(contact.metaData).toMutableMap()

                metaDataMap["host"] = host
                contact.metaData = converter.fromStringMap(metaDataMap)
            }

            if (name != null && contact.name != null) {
                if (name == contact.name && isVerified) {
                    contact.isVerified = true
                } else {
                    contact.isVerified = isVerified
                }
                contact.name = name
                saveContact(contact)
            }
        }
    }

    fun saveContact(contact: Contact) {
        val contactDao = App.instance.database!!.contactDao()
        contactDao.update(contact)
    }

    fun getAllRequests(): List<PayRequest>? {
        val payRequestDao = App.instance.database!!.payRequestDao()
        return payRequestDao.getAllPayRequests()
    }

    fun getPayRequest(accountID: HGCAccountID, amount: Long, name: String?, notes: String?): PayRequest? {
        val payRequestDao = App.instance.database!!.payRequestDao()
        val payRequests = payRequestDao.findPayRequest(accountID.stringRepresentation(), name!!, amount, notes!!)
        return if (payRequests != null && !payRequests.isEmpty()) {
            payRequests[0]
        } else null
    }

    fun createPayRequest(accountID: HGCAccountID, amount: Long, name: String?, notes: String?) {
        val payRequestDao = App.instance.database!!.payRequestDao()
        var payRequest = getPayRequest(accountID, amount, name, notes)
        if (payRequest == null) {
            payRequest = PayRequest(0, accountID.stringRepresentation(), name, notes, amount, Date())
            payRequestDao.insert(payRequest)

        } else {
            payRequest.importDate = Date()
            savePayRequest(payRequest)
        }
    }

    fun savePayRequest(request: PayRequest) {
        val payRequestDao = App.instance.database!!.payRequestDao()
        payRequestDao.update(request)
    }

    fun deletePayRequest(request: PayRequest) {
        val payRequestDao = App.instance.database!!.payRequestDao()
        payRequestDao.delete(request)
    }


    fun createTransaction(txn: Transaction, fromAccount: HGCAccountID?): TxnRecord {
        val dao = App.instance.database!!.txnRecordDao()
        var record = getTransaction(txn.getTxnBody().transactionID)
        if (record == null) {
            record = TxnRecord(txn.getTxnBody().transactionID)
            record.txn = txn.toByteArray()
            record.createdDate = Date()
            fromAccount?.let { record.fromAccId = it.stringRepresentation() }
            dao.insert(record)
            return dao.findRecordForTxnId(Converters().transactionIdToString(record.txnId)!!)[0]
        } else {
            return record
        }
    }

    fun createTransaction(transactionRecord: TransactionRecord): TxnRecord {
        val dao = App.instance.database!!.txnRecordDao()
        var record = getTransaction(transactionRecord.transactionID)
        if (record == null) {
            record = TxnRecord(transactionRecord.transactionID).apply {
                this.record = transactionRecord.toByteArray()
                createdDate = Date()
                this.parseProperties()
            }
            dao.insert(record)
            return dao.findRecordForTxnId(Converters().transactionIdToString(record.txnId)!!)[0]
        } else {
            return record
        }
    }

    fun getTransaction(txnId: TransactionID): TxnRecord? {
        val txnIdStr = Converters().transactionIdToString(txnId)
        val dao = App.instance.database!!.txnRecordDao()
        val list = dao.findRecordForTxnId(txnIdStr!!)
        return if (list != null && !list.isEmpty()) {
            list[0]
        } else null
    }

    fun updateTransaction(record: TxnRecord) {
        val dao = App.instance.database!!.txnRecordDao()
        dao.update(record)
    }

    fun getAllTxnRecord(fromAccount: Account?): List<TxnRecord> {

        val dao = App.instance.database!!.txnRecordDao()
        var records: List<TxnRecord>? = null
        if (fromAccount == null) {
            records = dao.getAllRecords()
        } else {
            if (fromAccount.accountID() != null) {
                records = dao.findRecordForAccountId(fromAccount.accountID()!!.stringRepresentation())
            } else {
                records = ArrayList()
            }
        }

        if (records != null && !records.isEmpty()) {
            val accounts = getAllAccounts()
            val myAccountIds = accounts.filter { it.accountID() != null }.map { it.accountID()!!.stringRepresentation() }
            for (record in records) {
                record.parseProperties()
                if (record.fromAccId != null) {
                    val account = getAccount(accounts, record.fromAccId)
                    if (account != null) {
                        record.fromAccount = account!!.getContact()
                    } else {
                        val contact = getContact(record.fromAccId!!)
                        if (contact != null) {
                            record.fromAccount = contact
                        }
                    }
                }

                if (record.toAccountId != null) {
                    val account = getAccount(accounts, record.toAccountId)
                    if (account != null) {
                        record.toAccount = account!!.getContact()
                    } else {
                        val contact = getContact(record.toAccountId!!)
                        if (contact != null) {
                            record.toAccount = contact
                        }
                    }
                    if (record.toAccountId != null) {
                        if (myAccountIds.contains(record.toAccountId!!)) {
                            record.isPositive = true
                        }
                    }
                }
            }
        }


        return records
    }

    fun deleteTxnRecord(record: TxnRecord) {
        val dao = App.instance.database!!.txnRecordDao()
        dao.delete(record)
    }

    private fun getAccount(accounts: List<Account>, accountId: String?): Account? {
        val accountID = HGCAccountID.fromString(accountId)
        if (accountID != null) {
            for (account in accounts) {
                val accId = account.accountID()
                if (accId != null && accId.stringRepresentation() == accountID.stringRepresentation()) {
                    return account
                }
            }
        }
        return null
    }

    fun getNode(host: String): Node? {
        val dao = App.instance.database!!.nodeDao()
        val list = dao.findNodeForHost(host)
        return if (list != null && !list.isEmpty()) {
            list[0]
        } else null
    }

    fun createNode(node: Node) {
        val dao = App.instance.database!!.nodeDao()
        val record = getNode(node.host!!)
        if (record == null) {
            dao.insert(node)
        }
    }

    fun updateNode(node: Node) {
        val dao = App.instance.database!!.nodeDao()
        dao.update(node)
    }

    fun getAllNodes(activeOnly: Boolean): List<Node> {
        val dao = App.instance.database!!.nodeDao()
        return if (activeOnly)
            dao.findActiveNodes()
        else
            dao.getAllNodes()
    }


    fun deleteAllNodes() {
        val dao = App.instance.database!!.nodeDao()
        dao.deleteAll()
    }

    fun updateWallet(wallet: Wallet?) {
        val walletDao = App.instance.database!!.walletDao()
        if (wallet != null)
            walletDao.update(wallet)

    }
}
